pkg/controller/master-controller-manager/cluster-template-synchronizer/controller.go

```diff
@@ -0,0 +1,196 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clustertemplatesynchronizer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	kubermaticapiv1 "k8c.io/kubermatic/v2/pkg/api/v1"
+	controllerutil "k8c.io/kubermatic/v2/pkg/controller/util"
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	kuberneteshelper "k8c.io/kubermatic/v2/pkg/kubernetes"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// This controller syncs the kubermatic cluster templates on the master cluster to the seed clusters.
+	ControllerName = "cluster_template_syncing_controller"
+)
+
+type reconciler struct {
+	log          *zap.SugaredLogger
+	masterClient ctrlruntimeclient.Client
+	seedClients  map[string]ctrlruntimeclient.Client
+	recorder     record.EventRecorder
+}
+
+func Add(
+	masterMgr manager.Manager,
+	seedManagers map[string]manager.Manager,
+	log *zap.SugaredLogger) error {
+
+	log = log.Named(ControllerName)
+	r := &reconciler{
+		log:          log,
+		masterClient: masterMgr.GetClient(),
+		seedClients:  map[string]ctrlruntimeclient.Client{},
+		recorder:     masterMgr.GetEventRecorderFor(ControllerName),
+	}
+
+	c, err := controller.New(ControllerName, masterMgr, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %v", err)
+	}
+
+	for seedName, seedManager := range seedManagers {
+		r.seedClients[seedName] = seedManager.GetClient()
+	}
+
+	// Watch for changes to ClusterTemplates
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.ClusterTemplate{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to watch cluster templates: %v", err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	log := r.log.With("resource", request.Name)
+	log.Debug("Processing")
+
+	err := r.reconcile(ctx, log, request)
+	if controllerutil.IsCacheNotStarted(err) {
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+	}
+	if err != nil {
+		log.Errorw("ReconcilingError", zap.Error(err))
+	}
+
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, request reconcile.Request) error {
+	clusterTemplate := &kubermaticv1.ClusterTemplate{}
+	if err := r.masterClient.Get(ctx, ctrlruntimeclient.ObjectKey{Name: request.Name}, clusterTemplate); err != nil {
+		if controllerutil.IsCacheNotStarted(err) {
```

Comment:
I guess we can remove this check because it will be returned in next `return ctrlruntimeclient.IgnoreNotFound(err)` anyways?